op-node/rollup/finality/finalizer.go

```diff
@@ -231,37 +235,49 @@ func (fi *Finalizer) tryFinalize() {
 	fi.mu.Lock()
 	defer fi.mu.Unlock()
 
-	gadgetActivatedTimestamp, err := fi.babylonFinalityClient.QueryBtcStakingActivatedTimestamp()
-	if err != nil && !strings.Contains(err.Error(), fgtypes.ErrBtcStakingNotActivated.Error()) {
-		fi.emitter.Emit(rollup.CriticalErrorEvent{Err: fmt.Errorf("failed to query BTC staking activated timestamp: %w", err)})
-		return
-	}
-
 	// overwritten if we finalize
 	finalizedL2 := fi.lastFinalizedL2 // may be zeroed if nothing was finalized since startup.
 	var finalizedDerivedFrom eth.BlockID
+
 	// go through the latest inclusion data, and find the last L2 block that was derived from a finalized L1 block
 	fi.log.Debug("try finalize", "finality_data", fi.finalityData, "last_finalized_l2", finalizedL2)
-	for _, fd := range fi.finalityData {
-		if fd.L2Block.Number > finalizedL2.Number && fd.L1Block.Number <= fi.finalizedL1.Number {
-			lastFinalizedBlock := fi.findLastBtcFinalizedL2Block(
-				fd.L2Block.Number, finalizedL2.Number, gadgetActivatedTimestamp)
-
-			// set finalized block(s)
-			if lastFinalizedBlock != nil {
-				finalizedL2 = *lastFinalizedBlock
+	if fi.babylonFinalityClient == nil {
```

Comment:
```suggestion
	if fi.EnableBTCStaking() {
```
and we can wrap this logic inside the `EnableBTCStaking` function

Reply:
Also what will happen if an OP node enables BTC staking and run OP node for a while, shuts down OP node, and restarts OP node with BTC staking disabled? Given that this function tries to write finalised block to DB, would it be possible that the finaliser needs to iterate over more blocks after the restart?

This seems to be another reason of keeping BTC stake finalised in a separate KV store, in parallel with unsafe/safe/finalised states of blocks in OP stack

Reply:
> what will happen if an OP node enables BTC staking and run OP node for a while, shuts down OP node, and restarts OP node with BTC staking disabled

in this case, once it's restarted, it will continue to operate normally (e.g. fetch L1 data, derive L2 blocks). when it tries to finalise a block, it will simply skip the BTC quorum check.

> would it be possible that the finaliser needs to iterate over more blocks after the restart

what do you mean by itereating over more blocks? after the restart, it won't call the gRPC of FG. it will become a vanilla op-node

> This seems to be another reason of keeping BTC stake finalised in a separate KV store, in parallel with unsafe/safe/finalised states of blocks in OP stack

this data is already stored in the FG KV db. `op-node` connects with it via gRPC

Reply:
> and we can wrap this logic inside the EnableBTCStaking function

@SebastianElvis do you mean to create a helper function like this?

```
func (fi *Finalizer) BTCStakingEnabled() bool {
	return fi.babylonFinalityClient != nil
}
```

this doesn't seem very useful

Reply:
> what do you mean by itereating over more blocks? after the restart, it won't call the gRPC of FG. it will become a vanilla op-node

Let's say block `h` is finalised in OP stack term + BTC stake finalised, and `h+1` to `h+5` are finalised in OP stack term but haven't been BTC stake finalised. Then upon retsart with BTC staking disabled, Op node starts iterating from `h+1` but not `h+5`. However it's supposed to start from `h+5`.

Reply:
got it. it's fine to re-derive from h1. although there will be some wasted rpc calls to L1, it doesn't require any changes in the codebase and such restart should be rare

Reply:
yep. not blocking for now but wdyt about having some doc / TODO for this?

Reply:
there isn't any TODO here as the optimization has low ROI. (i.e. it requires intrusive changes for marginally gain)

i'd say not adding any comments to avoid confusion